enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -128,154 +104,96 @@ New *pipeline* field:
+Elasticsearch is a special case; JSON records with _different formats_ must go to different indices, otherwise type conflicts and cardinality problems can occur.
+The elasticsearch output can be configured with a "structured type" that is used to construct an index name.
 
-New fields for output type `elasticsearch`
-* `indexKey`: (string, optional) Use value of meta-data key as `index` value, if present.
-  These keys are supported:
-  - `kubernetes.namespaceName`: Use the namespace name as the index name.
-  - `kubernetes.labels.<key>`: Use the string value of kubernetes label with key `<key>`.
-  - `openshift.labels.<key>`: Use the string value of an openshift label with key `<key>` (see [](forwarder-tagging.md))
-  - *other keys may be added in future*
-* `indexName`: (string, optional)
-  * If `indexKey` is not set, or the key is missing, use `indexName` as the `index` value.
-  * If `indexKey` is set and the key is present, that takes precedence over `indexName`
+New fields in  `output.elasticsearch`:
+* `structuredTypeName`: (string, optional) the structured type, unless `structuredTypeKey` is set, and the key is present.
+* `structuredTypeKey`: (string, optional) Use the value of this meta-data key (if present and non-empty)  as the structured type. These keys are supported:
+  * `kubernetes.labels.<key>`: Use the string value of kubernetes label with key `<key>`.
+  * `openshift.labels.<key>`: Use the string value of an openshift label with key `<key>` (see [forwarder-tagging.md]())
 
-**Note:** The Elasticsearch output will _delete_ the `structured` field if the rules above do not yield a non-empty index name.
-This is to avoid sending randomly structured JSON to shared indices.
+Notes:
+* The Elasticsearch _index_ for structured records is formed by appending "app-" to the structured type.
+* Unstructured records are not sent to the structured index, they are indexed as usual in application, infrastructure or audit indices.
+* If there is no non-empty structured type, forward an _unstructured_ record with no `structured` field.
 
-Indices are created automatically on-demand by co-operation between the forwarder and the managed default store.
+It is important not to overload elasticsearch with too many indices.
+Only use distinct structured types for distinct log _formats_, **not** for each application or namespace.
+For example, most Apache applications use the same JSON log format and should use the same structured type, for example "LogApache".
 
-#### Default output configuration
-
-To allow index settings on the default elasticsearch output, there is a new section in the ClusterLogForwarder spec:
-
-* outputDefaults (type map, optional):
-  Each map under outputDefaults MUST have a `type` field.
-  All other fields are used as defaults for any outputs of the same `type` _including_ the `default` output
+Structured indices are created automatically by the managed default store.
```

Comment:
we don't have to do anything in this proposal about it, but as an aside i suggest we start using "managed store" vs "default store" since its much more clear to everyone what that means (the alternative then would be "remote store")

Reply:
I use "default" in this proposal because I literally mean: the output named "default" in CLF configuration, which is also a managed ES instance.
Otherwise I agree.